### PR TITLE
JAMES-2366 REST API for Forwards CRUD

### PR DESCRIPTION
**Pull request by chibenwa**
Based on a WIP...

Only the 4 last commits:

 - https://github.com/linagora/james-project/commit/e5851242ef8ef1f9fcc518c8cb8fe6efb9b7e837 `
JAMES-2366 getAddress should return for forward `
 - https://github.com/linagora/james-project/commit/48b77bde73b4a65600b2994a159caeb3079583f5 `
JAMES-2366 Introduce REST API for forward `
 - https://github.com/linagora/james-project/commit/e92e20c79b9482a8bd7a60ff38d62f743874f55d `
JAMES-2366 Bind webadmin forward routes `
 - https://github.com/linagora/james-project/commit/374750465bb9aa2db5ffea8628dedffe8733d727 `
JAMES-2366 Adding documentation to forwards routes `

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[374750465bb9aa2db5ffea8628dedffe8733d727] [INFO] Apache James :: Server :: Data :: HBase Persistence  FAILURE [01:19 min]

[374750465bb9aa2db5ffea8628dedffe8733d727] Tests in error: 
[374750465bb9aa2db5ffea8628dedffe8733d727]   AbstractRecipientRewriteTableTest.addMappingShouldNotThrowWhenMappingAlreadyExistsWithAnOtherType:288->addMapping:338 » RecipientRewriteTable
[374750465bb9aa2db5ffea8628dedffe8733d727]   AbstractRecipientRewriteTableTest.removeForwardMappingShouldDelete:316->addMapping:344 » RecipientRewriteTable
[374750465bb9aa2db5ffea8628dedffe8733d727]   AbstractRecipientRewriteTableTest.testAliasDomainMapping:220 » NoClassDefFound
[374750465bb9aa2db5ffea8628dedffe8733d727]   AbstractRecipientRewriteTableTest.testRecursiveMapping:195 » NoClassDefFound c...
[374750465bb9aa2db5ffea8628dedffe8733d727]   AbstractRecipientRewriteTableTest.testStoreAndRetrieveAddressMapping:123 » NoClassDefFound
[374750465bb9aa2db5ffea8628dedffe8733d727]   AbstractRecipientRewriteTableTest.testStoreAndRetrieveErrorMapping:145 » NoClassDefFound
[374750465bb9aa2db5ffea8628dedffe8733d727]   AbstractRecipientRewriteTableTest.testStoreAndRetrieveRegexMapping:95 » NoClassDefFound
```

test this please

**Comment by chibenwa** [comment]
Exact error is:

```

[374750465bb9aa2db5ffea8628dedffe8733d727] java.lang.NoClassDefFoundError: com/google/common/base/MoreObjects
[374750465bb9aa2db5ffea8628dedffe8733d727] 	at org.apache.james.rrt.lib.AbstractRecipientRewriteTableTest.testAliasDomainMapping(AbstractRecipientRewriteTableTest.java:220)
[374750465bb9aa2db5ffea8628dedffe8733d727] Caused by: java.lang.ClassNotFoundException: com.google.common.base.MoreObjects
[374750465bb9aa2db5ffea8628dedffe8733d727] 	at org.apache.james.rrt.lib.AbstractRecipientRewriteTableTest.testAliasDomainMapping(AbstractRecipientRewriteTableTest.java:220)
```

I propose to fix that in another PR...

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+    private void createServer(ForwardRoutes forwardRoutes) throws Exception {
+        webAdminServer = WebAdminUtils.createWebAdminServer(
+            new DefaultMetricFactory(),
+            forwardRoutes);
+        webAdminServer.configure(NO_CONFIGURATION);
+        webAdminServer.await();
+
+        RestAssured.requestSpecification = WebAdminUtils.buildRequestSpecification(webAdminServer)
+            .setBasePath(ForwardRoutes.ROOT_PATH)
+            .log(LogDetail.ALL)
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java]: we may not want to log everything

**Comment by chibenwa** [comment]
> You didn't wrote any integration tests ? Is it planned for a later PR ?

@mbaechler  yes. This changeset only expose a CRUD for `Forwards`, and is indepandant from most of https://github.com/linagora/james-project/pull/1348 (which introduce and handle forward prefix). The forward is stored here, but corresponding address rewritting logic not applied.

An integration test will be done on top of both PR, as planned with @aduprat  on the team chat.

File: src/site/markdown/server/manage-webadmin.md [hunk 2]
```diff
@@ -837,6 +839,89 @@ Response codes:
  - 400: Group structure or member is not valid
  - 500: Internal error
 
+## Creating address forwars
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: Good catch.

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+
+    private void createServer(ForwardRoutes forwardRoutes) throws Exception {
+        webAdminServer = WebAdminUtils.createWebAdminServer(
+            new DefaultMetricFactory(),
+            forwardRoutes);
+        webAdminServer.configure(NO_CONFIGURATION);
+        webAdminServer.await();
+
+        RestAssured.requestSpecification = WebAdminUtils.buildRequestSpecification(webAdminServer)
+            .setBasePath(ForwardRoutes.ROOT_PATH)
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java]: Funny, I copied one of your test class, on Groups. 

I will also sanitize that class, then...

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+            .then()
+                .contentType(ContentType.JSON)
+                .statusCode(HttpStatus.OK_200)
+                .body(is("[]"));
+        }
+
+        @Test
+        void getForwardShouldListExistingForwardsInOrder() {
+            given()
+                .put(ALICE + SEPARATOR + BOB);
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java]: Funny, I transposed the one for groups (as the functional needs are the same).

I think it's a good idea as it allows better extensibility, of course.

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java [hunk 5]
```diff
@@ -0,0 +1,715 @@
...
+                when()
+                    .get(ALICE)
+                .then()
+                    .contentType(ContentType.JSON)
+                    .statusCode(HttpStatus.OK_200)
+                    .extract()
+                    .body()
+                    .jsonPath()
+                    .getList(".");
+            assertThat(addresses).containsExactly(BOB);
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java]: Idem I 100% agree, but `Groups` might also benefit from this change. WDYT?

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+                    .body()
+                    .jsonPath()
+                    .getList(".");
+            assertThat(addresses).containsExactly(BOB, CEDRIC);
+        }
+
+        @Test
+        void putUserInForwardShouldAllowBaseBeingADestination() {
+            given()
+                .put(ALICE + SEPARATOR + ALICE);
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/ForwardRoutesTest.java]: WDYT of `forwardShouldAllowIdentity` ?

**Comment by chibenwa** [comment]
Rebased and force pushed in order to solve conflicts.

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/1364

**Pull request closed without merging**